certbot-apache/certbot_apache/dualparser.py

```diff
@@ -0,0 +1,266 @@
+""" Tests for ParserNode interface """
+from certbot_apache import assertions
+from certbot_apache import augeasparser
+
+
+class DualNodeBase(assertions.InterfaceAssertions):
+    """ Dual parser interface for in development testing. This is used as the
+    base class for dual parser interface classes. This class handles runtime
+    attribute value assertions."""
+
+    def save(self, msg):
+        """ Call save for both parsers """
+        self.primary.save(msg)
+        self.secondary.save(msg)
+
+    def __getattr__(self, aname):
+        """ Attribute value assertion """
+        firstval = getattr(self.primary, aname)
+        secondval = getattr(self.secondary, aname)
+        if not self.isPass(firstval, secondval):
+            self.assertSimple(firstval, secondval)
+        return firstval
+
+
+class DualCommentNode(DualNodeBase):
+    """ Dual parser implementation of CommentNode interface """
+
+    def __init__(self, comment="", ancestor=None, filepath="", dirty=False,
+                 primary=None, secondary=None):
+        """ This initialization implementation allows ordinary initialization
+        of CommentNode objects as well as creating a DualCommentNode object
+        using precreated or fetched CommentNode objects if provided as optional
+        arguments primary and secondary.
+        """
+
+        if not primary:
+            self.primary = augeasparser.AugeasCommentNode(comment, ancestor,
+                                                          filepath, dirty)
+        else:
+            self.primary = primary
+        if not secondary:
+            self.secondary = augeasparser.AugeasCommentNode(comment, ancestor,
+                                                            filepath, dirty)
+        else:
+            self.secondary = secondary
```

Comment:
I think at the bottom of these `__init__` methods we should also assert equality and doing so simplifies some of the code elsewhere such as creating children.